lib/core/src/Cardano/DB/Sqlite/Delete.hs

```diff
@@ -0,0 +1,237 @@
+--
+-- We can't really guarantee that these are all off, so what we can do is
+-- whenever after an rm the file still exists to try again and wait a bit.
+--
+-- See <https://github.com/haskell/directory/issues/96> for more information
+-- about this issue.
+deleteSqliteDatabase :: Tracer IO DeleteSqliteDatabaseLog -> FilePath -> IO ()
+deleteSqliteDatabase tr =
+    deleteSqliteDatabase' tr deleteSqliteDatabaseRetryPolicy
+
+-- | A variant of 'deleteSqliteDatabase' where the caller can specify the
+-- 'RetryPolicy'.
+deleteSqliteDatabase'
+    :: Tracer IO DeleteSqliteDatabaseLog
+    -> RetryPolicy
+    -> FilePath
+    -> IO ()
+deleteSqliteDatabase' tr pol db = mapM_ delete files
+  where
+    files = [ db, db <> "-wal", db <> "-shm" ]
+    delete = handleErrors tr pol . removePathForcibly
+
+handleErrors
+    :: Tracer IO DeleteSqliteDatabaseLog
+    -> RetryPolicy
+    -> IO ()
+    -> IO ()
+#if defined(mingw32_HOST_OS)
+handleErrors tr policy = recovering policy [check] . const
+  where
+    check = logRetries (pure . isPermissionError) logRetry
+    logRetry True _ st = traceWith tr $ MsgRetryDelete $ rsIterNumber st
+    logRetry False e _ = traceWith tr $ MsgGaveUpDelete $ show e
+#else
+handleErrors _ _ = id
+#endif
+
+-- | Retry policy where delay increases linearly from base with each retry.
+-- (as <https://www.sqlite.org/src/info/89f1848d7f implemented by SQLite>)
+linearBackoff
+    :: Int
+    -- ^ Base delay in microseconds
+    -> RetryPolicy
+linearBackoff base = retryPolicy $ \ RetryStatus { rsIterNumber = n } ->
+  Just $! base * n
+
+-- | Recommended retry policy for 'deleteSqliteDatabase'.
+deleteSqliteDatabaseRetryPolicy :: RetryPolicy
+deleteSqliteDatabaseRetryPolicy = linearBackoff 25000 <> limitRetries 10
+
+-- | Log messages that may arise from 'deleteSqliteDatabase'.
+data DeleteSqliteDatabaseLog
+    = MsgRetryDelete Int
+    | MsgGaveUpDelete String
+    deriving (Generic, Show, Eq, ToJSON)
+
+instance ToText DeleteSqliteDatabaseLog where
+    toText msg = case msg of
+        MsgRetryDelete retryNum ->
+            "retry " <> T.pack (show retryNum) <> " for lock/sharing violation - probably due to antivirus software"
+        MsgGaveUpDelete e ->
+            "gave up on delete due to " <> T.pack e
+
+instance DefinePrivacyAnnotation DeleteSqliteDatabaseLog
+instance DefineSeverity DeleteSqliteDatabaseLog where
+    defineSeverity msg = case msg of
+        MsgRetryDelete _ -> Warning
+        MsgGaveUpDelete _ -> Error
+
+{-------------------------------------------------------------------------------
+                          Ref-counting open databases
+-------------------------------------------------------------------------------}
+
+-- | Mutable variable containing reference counts to IDs of type @ix@.
+data RefCount ix = RefCount
+    { _refCount :: MVar (Map ix Int) -- ^ number of references to each index
+    , _takeLock :: MVar () -- ^ lock on incrementing references
+    }
```

Comment:
I don't think we do have more than one `withDatabase` call per wallet, so there should be always at most a single connection per database :thinking:. 
If we can do more, we can also do less so I guess it's fine.

Reply:
I thought that might be the case, but am hoping it will be more resilient if it supports more than one `withDatabase` call per wallet.